### ISSUE TITLE
modeling/embedding_lookup function issue

### ISSUE THREAD

**Issue by zqcchris** [google-research/bert#294]
def embedding_lookup(input_ids, vocab_size, embedding_size=128, initializer_range=0.02,
                     word_embedding_name="word_embeddings", use_one_hot_embeddings=False):
    """Looks up words embeddings for id tensor.

  Args:
    input_ids: int32 Tensor of shape [batch_size, seq_length] containing word
      ids.
    vocab_size: int. Size of the embedding vocabulary.
    embedding_size: int. Width of the word embeddings.
    initializer_range: float. Embedding initialization range.
    word_embedding_name: string. Name of the embedding table.
    use_one_hot_embeddings: bool. If True, use one-hot method for word
      embeddings. If False, use `tf.nn.embedding_lookup()`. One hot is better
      for TPUs.

  Returns:
    float Tensor of shape [batch_size, seq_length, embedding_size].
  """
    # This function assumes that the input is of shape [batch_size, seq_length,
    # num_inputs].
    #
    # If the input is a 2D tensor of shape [batch_size, seq_length], we
    # reshape to [batch_size, seq_length, 1].
    # if input_ids.shape.ndims == 2:
    #     input_ids = tf.expand_dims(input=input_ids, axis=[-1])

    embedding_table = tf.get_variable(name=word_embedding_name, shape=[vocab_size, embedding_size],
                                      initializer=create_initializer(initializer_range))

    if use_one_hot_embeddings:
        flat_input_ids = tf.reshape(tensor=input_ids, shape=[-1])
        one_hot_input_ids = tf.one_hot(indices=flat_input_ids, depth=vocab_size)
        output = tf.matmul(one_hot_input_ids, embedding_table)
    else:
        output = tf.nn.embedding_lookup(params=embedding_table, ids=input_ids)

    # input_shape = get_shape_list(input_ids)
    # output = tf.reshape(tensor=output, shape=input_shape[0:-1] + [input_shape[-1] * embedding_size])
    return output, embedding_table
![default](https://user-images.githubusercontent.com/23413723/50369730-06b1cf80-05d5-11e9-957e-fc7b84460830.PNG)

there is something weird in the embedding_lookup function. In the start of this function, the input_ids dimension has been expanded if its ndims==2, however after embedding_lookup operation,   we reshape the output. I think these two operation hedged with each other.

In another way to illustrate this issue, I suppose that the shape of imput_ids is [1, 128]
after dimension expanded, the shape of input_ids is [1, 128, 1]
after embedding_lookup, the shape has been changed into [1, 128, 1, 768]  # bmedding_size = 768
then we reshape the output of previous operation to [1, 128, 768]
so why don't we cancel the first expand opetaion, and we can get the embedding_lookup output with shape [1, 128, 768]
